File: problems/01/two_sum.py
def two_sum(nums: list[int], target: int) -> list[int]:
    table: dict[int, int] = {}
    for i, num in enumerate(nums):
        compliment = target - num
        if compliment in table:
            return [table[compliment], i]
        table[num] = i

    return []


def test_two_sum():
    assert two_sum([2, 7, 11, 15], 9) == [0, 1]
    assert two_sum([3, 2, 4], 6) == [1, 2]
    assert two_sum([3, 3], 6) == [0, 1]


if __name__ == "__main__":
    import pytest

    pytest.main(["-xvv", __file__])

File: problems/01/two_sum.exs
defmodule TwoSum do
  def iterative(nums, target) do
    nums
    |> Enum.with_index()
    |> Enum.reduce_while(%{}, fn {num, idx}, acc ->
      complement = Map.get(acc, target - num)

      if complement do
        {:halt, [complement, idx]}
      else
        {:cont, Map.put(acc, num, idx)}
      end
    end)
  end

  def recursive(nums, target, cache \\ %{}, index \\ 0) do
    [num | rest] = nums
    comp = target - num

    case Map.get(cache, comp) do
      nil -> recursive(rest, target, Map.put(cache, num, index), index + 1)
      comp_index -> [comp_index, index]
    end
  end
end

defmodule TwoSumTest do
  use ExUnit.Case

  test "leetcode examples" do
    assert TwoSum.iterative([2, 7, 11, 15], 9) == [0, 1]
    assert TwoSum.recursive([2, 7, 11, 15], 9) == [0, 1]
    assert TwoSum.iterative([3, 2, 4], 6) == [1, 2]
    assert TwoSum.recursive([3, 2, 4], 6) == [1, 2]
    assert TwoSum.iterative([3, 3], 6) == [0, 1]
    assert TwoSum.recursive([3, 3], 6) == [0, 1]
  end
end
